common/channel/src/libra_channel.rs

```diff
@@ -0,0 +1,151 @@
+/// channel so that we can ensure fairness among validators
+pub trait ValidatorMessage {
+    /// Extract the Validator from which this message arrived. This
+    /// will be used for ensuring fairness among validators.
+    fn get_validator(&self) -> &str;
+}
+
+/// Type is an enum which can be used as a configuration option for
+/// libra_channel
+pub enum Type {
+    LIFO,
+    FIFO,
+}
+
+struct PerValidatorQueue<T> {
+    queue_type: Type,
+    per_validator_queue: HashMap<String, VecDeque<T>>,
+    max_queue_size: usize,
+}
+
+impl<T: ValidatorMessage> PerValidatorQueue<T> {
+    fn new(queue_type: Type, max_queue_size: usize) -> Self {
+        assert!(
+            max_queue_size > 0,
+            "cannot have a zero-capacity libra_channel"
+        );
+        Self {
+            queue_type,
+            per_validator_queue: HashMap::new(),
+            max_queue_size,
+        }
+    }
+
+    fn add_message(&mut self, message: T) {
+        let validator = message.get_validator();
+        if !self.per_validator_queue.contains_key(validator) {
+            self.per_validator_queue.insert(
+                validator.to_string(),
+                VecDeque::with_capacity(self.max_queue_size + 1),
+            );
+        }
+        if let Some(q) = self.per_validator_queue.get_mut(validator) {
+            q.push_back(message);
+            if q.len() == self.max_queue_size + 1 {
+                match self.queue_type {
+                    // Drop old message for FIFO
+                    Type::FIFO => q.pop_front(),
+                    // Drop new message for LIFO
+                    Type::LIFO => q.pop_back(),
+                };
+            }
+        }
+    }
+
+    // TODO: Come up with a better data structure to efficiently
+    // do a round robin among validators
+    fn get_message(&mut self) -> Option<T> {
+        for q in self.per_validator_queue.values_mut() {
+            if !q.is_empty() {
+                return match self.queue_type {
+                    Type::FIFO => q.pop_back(),
+                    Type::LIFO => q.pop_front(),
+                };
+            }
+        }
+        None
+    }
+}
+
+struct SharedState<T> {
+    /// The data stored by this Channel
+    /// TODO This needs to be updated to a data structure which can do round robin for validators
+    data: PerValidatorQueue<T>,
+
+    /// The thread can use this after calling `put` to tell
+    /// `Receiver`'s task to wake up, and process the next item
+    waker: Option<Waker>,
+}
+
+/// The sending end of the Channel
+#[derive(Clone)]
+pub struct Sender<T: ValidatorMessage> {
+    shared_state: Arc<Mutex<SharedState<T>>>,
+}
+
+/// The receiving end of the Channel
+pub struct Receiver<T: ValidatorMessage> {
+    shared_state: Arc<Mutex<SharedState<T>>>,
+}
+
+impl<T: ValidatorMessage> Sender<T> {
+    /// TODO: We can have this return the old data if the new data is replacing some existing data
```

Comment:
Yup, although it'd then depend on FIFO / LIFO ordering (in FIFO ordering the returned value would be the message itself?)

Reply:
True. Thinking some more about it, the actual message would not be of any use if we using it for back-pressure. Only a boolean value is enough to indicate that the queue is full.